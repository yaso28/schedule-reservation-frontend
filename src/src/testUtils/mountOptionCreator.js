import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
