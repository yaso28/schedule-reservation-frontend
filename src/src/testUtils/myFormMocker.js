import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
