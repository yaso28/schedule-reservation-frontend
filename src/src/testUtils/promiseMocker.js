import _ from 'lodash'

const createSetTimeoutPromise = getResult => new Promise((resolve, reject) => {
  setTimeout(() => getResult(resolve, reject), 1000)
})
const createSwitchPromise = (value, error, switchObj) => createSetTimeoutPromise(
  (resolve, reject) => switchObj.isSuccess ? resolve(value) : reject(error)
)
const createResolvedPromise = value => createSetTimeoutPromise(resolve => resolve(value))
const createRejectedPromise = error => createSetTimeoutPromise((resolve, reject) => reject(error))

const createMock = createPromise => jest.fn().mockImplementationOnce(createPromise)
const createMockForCases = (caseList, createPromiseForCurrentCase) => jest.fn((...args) => {
  for (let i = 0; i < caseList.length; i++) {
    const currentCase = caseList[i]
    if (_.isEqual(args, currentCase.args)) {
      return createPromiseForCurrentCase(currentCase)
    }
  }
  throw new Error('createMockForCases: no args matched')
})

class PromiseMocker {
  init () {
    jest.useFakeTimers()
  }

  clear () {
    jest.clearAllTimers()
  }

  flushPendings () {
    jest.runOnlyPendingTimers()
  }

  createSwitch () {
    return { isSuccess: true }
  }

  setSwitch (switchObj, isSuccess) {
    switchObj.isSuccess = isSuccess
  }

  mock (value, error, switchObj) {
    return createMock(() => createSwitchPromise(value, error, switchObj))
  }

  mockResolved (value) {
    return createMock(() => createResolvedPromise(value))
  }

  mockRejected (error) {
    return createMock(() => createRejectedPromise(error))
  }

  mockForCases (caseList) {
    return createMockForCases(caseList, currentCase => createSwitchPromise(currentCase.value, currentCase.error, currentCase.switch))
  }

  mockResolvedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createResolvedPromise(currentCase.value))
  }

  mockRejectedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createRejectedPromise(currentCase.error))
  }
}

export default new PromiseMocker()
