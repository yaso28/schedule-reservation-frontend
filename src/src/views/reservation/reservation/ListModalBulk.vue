<template>
  <CModal
    :show="visible"
    :title="$t('action.bulk_change')"
    color="primary"
    centered
    @update:show="updateShow"
  >
    <CButton
      class="show-target"
      color="info"
      variant="outline"
      @click="toggleTarget"
    >
      {{ $t('action.bulk_change') }}{{ $t('common.target') }}
    </CButton>
    <CCollapse
      :show="showTarget"
    >
      <CDataTable
        v-if="items"
        :items="items"
        :fields="fields"
        :header="false"
        add-table-classes="mt-2 mb-0 text-nowrap"
      >
        <template #ymd="{item}">
          <td>
            <MyDateSpan :value="item.schedule.ymd" />
          </td>
        </template>
        <template #time="{item}">
          <td>
            {{ item.begins_at }} - {{ item.ends_at }}
          </td>
        </template>
        <template #schedule_place="{item}">
          <td>
            {{ item.schedule.schedule_place.name }}
          </td>
        </template>
        <template #reservation_status="{item}">
          <td>
            {{ item.reservation_status.name }}
          </td>
        </template>
      </CDataTable>
      <CButton
        class="hide-target"
        color="info"
        variant="outline"
        @click="toggleTarget"
      >
        {{ $t('action.hide') }}
      </CButton>
    </CCollapse>

    <hr class="mt-2 mb-3">

    <div v-if="form">
      <MySelect
        v-if="reservationStatusList"
        :label="$t('reservation_status.model')"
        :form="form"
        field="reservation_status_id"
        :options="reservationStatusList"
      />
    </div>
    <template #footer>
      <CButton
        class="btn-save"
        type="submit"
        color="primary"
        @click="save"
      >
        {{ $t('action.save') }}
      </CButton>
    </template>
  </CModal>
</template>

<script>
export default {
  name: 'ListModalBulk',
  components: {
    MyDateSpan: require('@/components/MyDateSpan').default,
    MySelect: require('@/components/inputs/MySelect').default
  },
  inheritAttrs: false,
  inject: ['apiService'],
  props: {
    show: {
      type: Boolean,
      default: false
    },
    items: {
      type: Array,
      default: () => []
    },
    reservationStatusList: {
      type: Array,
      default: () => []
    }
  },
  data () {
    return {
      visible: this.show,
      form: null,
      showTarget: false
    }
  },
  computed: {
    fields () {
      return [
        { key: 'ymd', label: this.$t('schedule.ymd') },
        { key: 'time', label: this.$t('schedule.time') },
        { key: 'schedule_place', label: this.$t('schedule_place.model') },
        { key: 'reservation_status', label: this.$t('reservation_status.model') }
      ]
    }
  },
  watch: {
    show (to) {
      if (to) {
        this.onShow()
      } else {
        this.onHide()
      }
    }
  },
  mounted () {
    if (this.show) {
      this.onShow()
    }
  },
  methods: {
    updateShow (value) {
      this.$emit('update:show', value)
    },
    onShow () {
      if (!this.items || !this.reservationStatusList) {
        this.updateShow(false)
        return
      }
      this.form = this.apiService.makeForm({
        id_list: this.items.map(item => item.id),
        reservation_status_id: null
      })
      this.showTarget = this.items.length <= 5
      this.visible = true
    },
    onHide () {
      this.visible = false
      this.showTarget = false
    },
    toggleTarget () {
      this.showTarget = !this.showTarget
    },
    async save () {
      try {
        this.$store.commit('alert/clear')
        this.$store.commit('processing/beginPost')
        await this.apiService.bulkChangeReservation(this.form)
        this.$store.dispatch('alert/saveSuccessNoRedirect')
        this.updateShow(false)
        this.$emit('save')
      } catch (e) {
        if (!this.apiService.isHandledError(e)) {
          throw e
        }
      } finally {
        this.$store.commit('processing/endPost')
      }
    }
  }
}
</script>
