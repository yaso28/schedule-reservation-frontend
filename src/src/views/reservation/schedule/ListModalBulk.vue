<template>
  <CModal
    :show="visible"
    :title="$t('action.bulk_change')"
    color="primary"
    centered
    @update:show="updateShow"
  >
    <CButton
      class="show-target"
      color="info"
      variant="outline"
      @click="toggleTarget"
    >
      {{ $t('action.bulk_change') }}{{ $t('common.target') }}
    </CButton>
    <CCollapse
      :show="showTarget"
    >
      <CDataTable
        v-if="items"
        :items="items"
        :fields="fields"
        :header="false"
        add-table-classes="mt-2 mb-0 text-nowrap"
      >
        <template #ymd="{item}">
          <td>
            <MyDateSpan :value="item.ymd" />
          </td>
        </template>
        <template #time="{item}">
          <td>
            {{ item.begins_at }} - {{ item.ends_at }}
          </td>
        </template>
        <template #schedule_place="{item}">
          <td>
            {{ item.schedule_place.name }}
          </td>
        </template>
        <template #schedule_usage="{item}">
          <td>
            {{ item.schedule_usage.name }}
          </td>
        </template>
        <template #reservation_status="{item}">
          <td>
            {{ item.reservation_status.name }}
          </td>
        </template>
        <template #schedule_status="{item}">
          <td>
            {{ item.schedule_status.name }}
          </td>
        </template>
      </CDataTable>
      <CButton
        class="hide-target"
        color="info"
        variant="outline"
        @click="toggleTarget"
      >
        {{ $t('action.hide') }}
      </CButton>
    </CCollapse>

    <hr class="mt-2 mb-3">

    <div v-if="form">
      <MySelect
        v-if="scheduleStatusList"
        :label="$t('schedule_status.model')"
        :form="form"
        field="schedule_status_id"
        :options="scheduleStatusList"
      />
    </div>
    <template #footer>
      <CButton
        class="btn-save"
        type="submit"
        color="primary"
        @click="save"
      >
        {{ $t('action.save') }}
      </CButton>
    </template>
  </CModal>
</template>

<script>
export default {
  name: 'ListModalBulk',
  components: {
    MyDateSpan: require('@/components/MyDateSpan').default,
    MySelect: require('@/components/inputs/MySelect').default
  },
  inheritAttrs: false,
  inject: ['apiService'],
  props: {
    show: {
      type: Boolean,
      default: false
    },
    items: {
      type: Array,
      default: () => []
    },
    scheduleStatusList: {
      type: Array,
      default: () => []
    }
  },
  data () {
    return {
      visible: this.show,
      form: null,
      showTarget: false
    }
  },
  computed: {
    fields () {
      return [
        { key: 'ymd', label: this.$t('schedule.ymd') },
        { key: 'time', label: this.$t('schedule.time') },
        { key: 'schedule_place', label: this.$t('schedule_place.model') },
        { key: 'schedule_usage', label: this.$t('schedule_usage.model') },
        { key: 'reservation_status', label: this.$t('reservation_status.model') },
        { key: 'schedule_status', label: this.$t('schedule_status.model') }
      ]
    }
  },
  watch: {
    show (to) {
      if (to) {
        this.onShow()
      } else {
        this.onHide()
      }
    }
  },
  mounted () {
    if (this.show) {
      this.onShow()
    }
  },
  methods: {
    updateShow (value) {
      this.$emit('update:show', value)
    },
    onShow () {
      if (!this.items || !this.scheduleStatusList) {
        this.updateShow(false)
        return
      }
      this.form = this.apiService.makeForm({
        id_list: this.items.map(item => item.id),
        schedule_status_id: null
      })
      this.showTarget = this.items.length <= 5
      this.visible = true
    },
    onHide () {
      this.visible = false
      this.showTarget = false
    },
    toggleTarget () {
      this.showTarget = !this.showTarget
    },
    async save () {
      try {
        this.$store.commit('alert/clear')
        this.$store.commit('processing/beginPost')
        await this.apiService.bulkChangeSchedule(this.form)
        this.$store.dispatch('alert/saveSuccessNoRedirect')
        this.updateShow(false)
        this.$emit('save')
      } catch (e) {
        if (!this.apiService.isHandledError(e)) {
          throw e
        }
      } finally {
        this.$store.commit('processing/endPost')
      }
    }
  }
}
</script>
