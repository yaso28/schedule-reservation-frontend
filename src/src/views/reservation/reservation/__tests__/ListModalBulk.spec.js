import ListModalBulk from '../ListModalBulk'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import promiseMocker from '@/testUtils/promiseMocker'
import storeModules from '@/store/modules'
import myFormMocker from '@/testUtils/myFormMocker'
import { mount } from '@vue/test-utils'

jest.mock('@/store/modules')
jest.mock('@/models/MyForm')

const reservationStatusList = require('@/testUtils/dummyApiData/reservation-status/list.json')
const reservationList = require('@/testUtils/dummyApiData/reservation/list.json')

const getBtnSaveWrapper = wrapper => wrapper.find('button.btn-save')

const assertClose = function (wrapper) {
  expect(wrapper.emitted()['update:show'][0][0]).toBe(false)
}
const assertNotClose = function (wrapper) {
  expect(wrapper.emitted()['update:show']).toBeFalsy()
}
const getSaveEventLog = wrapper => wrapper.emitted().save

describe('ListModalBulk', () => {
  let mountOption
  let apiService
  beforeEach(() => {
    promiseMocker.init()
    mountOption = mountOptionCreator.create(storeModules)
    mountOption.propsData = {
      show: false,
      items: null,
      reservationStatusList
    }
    apiService = mountOption.provide.apiService
  })

  afterEach(() => {
    promiseMocker.clear()
    jest.resetAllMocks()
  })

  describe('マウント', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = mount(ListModalBulk, mountOption)
    })

    describe('items：なし', () => {
      describe('表示', () => {
        beforeEach(async () => {
          apiService.makeForm = jest.fn()
          await wrapper.setProps({ show: true })
        })

        test('閉じる', () => {
          assertClose(wrapper)
        })

        test('フォームは作成しない', () => {
          expect(apiService.makeForm).not.toHaveBeenCalled()
        })
      })
    })

    describe('items：あり', () => {
      let items
      beforeEach(async () => {
        items = reservationList.slice(1, 4)
        await wrapper.setProps({ items })
      })

      describe('表示', () => {
        let form
        let makeFormArgs
        beforeEach(async () => {
          makeFormArgs = {
            id_list: items.map(item => item.id),
            reservation_status_id: null
          }
          form = myFormMocker.createMock(makeFormArgs)
          apiService.makeForm = jest.fn().mockReturnValueOnce(form)
          await wrapper.setProps({ show: true })
        })

        test('閉じない', () => {
          assertNotClose(wrapper)
        })

        test('フォーム作成', () => {
          expect(apiService.makeForm).toHaveBeenCalledWith(makeFormArgs)
        })

        test('モーダル全体表示', () => {
          expect(wrapper.element).toMatchSnapshot()
        })

        describe('保存ボタンクリック', () => {
          let saveError
          let saveSwitch
          beforeEach(async () => {
            saveError = new Error('save')
            saveSwitch = promiseMocker.createSwitch()
            apiService.bulkChangeReservation = promiseMocker.mock(null, saveError, saveSwitch)
            await getBtnSaveWrapper(wrapper).trigger('click')
          })

          test('アラートクリア', () => {
            expect(storeModules.alert.mutations.clear).toHaveBeenCalled()
          })

          test('Processingカバー表示開始', () => {
            expect(storeModules.processing.mutations.beginPost).toHaveBeenCalled()
          })

          test('保存処理を呼出', () => {
            expect(apiService.bulkChangeReservation).toHaveBeenCalledWith(form)
          })

          test('Processingカバー表示中', () => {
            expect(storeModules.processing.mutations.endPost).not.toHaveBeenCalled()
          })

          describe('保存：エラー', () => {
            beforeEach(() => {
              promiseMocker.setSwitch(saveSwitch, false)
              apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
              promiseMocker.flushPendings()
            })

            test('エラー処理', () => {
              expect(apiService.isHandledError).toHaveBeenCalledWith(saveError)
            })

            test('成功アラートは表示しない', () => {
              expect(storeModules.alert.actions.saveSuccessNoRedirect).not.toHaveBeenCalled()
            })

            test('閉じない', () => {
              assertNotClose(wrapper)
            })

            test('保存成功イベントは投げない', () => {
              expect(getSaveEventLog(wrapper)).toBeFalsy()
            })

            test('Processingカバー表示終了', () => {
              expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
            })
          })

          describe('保存：成功', () => {
            beforeEach(() => {
              promiseMocker.flushPendings()
            })

            test('成功アラート表示', () => {
              expect(storeModules.alert.actions.saveSuccessNoRedirect).toHaveBeenCalled()
            })

            test('閉じる', () => {
              assertClose(wrapper)
            })

            test('保存成功イベント投げる', () => {
              expect(getSaveEventLog(wrapper)).toBeTruthy()
            })

            test('Processingカバー表示終了', () => {
              expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
            })
          })
        })
      })
    })
  })
})
