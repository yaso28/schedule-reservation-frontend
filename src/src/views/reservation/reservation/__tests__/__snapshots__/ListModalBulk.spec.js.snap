// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListModalBulk マウント items：あり 表示 モーダル全体表示 1`] = `
<div>
  <div
    class="modal overflow-auto fade d-block modal-primary"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
      role="document"
    >
      <div
        class="modal-content"
      >
        <header
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
             一括変更 
          </h5>
          <button
            aria-label="Close"
            class="close"
            type="button"
          >
            ×
          </button>
        </header>
        <div
          class="modal-body"
        >
          <button
            class="btn show-target btn-outline-info"
            type="button"
          >
            
    一括変更対象
  
          </button>
           
          <div
            class=""
            style="height: 0px; overflow: hidden; transition: all 400ms ease-in-out;"
          >
            <div
              data-v-5f7284b9=""
            >
              <!---->
              <div
                class="position-relative table-responsive"
                data-v-5f7284b9=""
              >
                <table
                  class="table mt-2 mb-0 text-nowrap"
                  data-v-5f7284b9=""
                >
                  <thead
                    data-v-5f7284b9=""
                  >
                    <!---->
                    <!---->
                  </thead>
                  <tbody
                    class="position-relative"
                    data-v-5f7284b9=""
                  >
                    <tr
                      data-v-5f7284b9=""
                    >
                      <td
                        data-v-5f7284b9=""
                      >
                        <span
                          data-v-5f7284b9=""
                        >
                          
  2020/12/05
  
                          <span
                            class="text-primary"
                          >
                            
    (土)
  
                          </span>
                        </span>
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          10:00 - 13:00
        
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          多目的ホール＠公民館1F
        
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          書類控取得
        
                      </td>
                    </tr>
                    <!---->
                    <tr
                      data-v-5f7284b9=""
                    >
                      <td
                        data-v-5f7284b9=""
                      >
                        <span
                          data-v-5f7284b9=""
                        >
                          
  2020/12/05
  
                          <span
                            class="text-primary"
                          >
                            
    (土)
  
                          </span>
                        </span>
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          13:00 - 16:00
        
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          体育館＠公民館別棟
        
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          書類控取得
        
                      </td>
                    </tr>
                    <!---->
                    <tr
                      data-v-5f7284b9=""
                    >
                      <td
                        data-v-5f7284b9=""
                      >
                        <span
                          data-v-5f7284b9=""
                        >
                          
  2020/12/06
  
                          <span
                            class="text-danger"
                          >
                            
    (日)
  
                          </span>
                        </span>
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          10:00 - 13:00
        
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          体育館＠公民館別棟
        
                      </td>
                      <td
                        data-v-5f7284b9=""
                      >
                        
          キャンセル
        
                      </td>
                    </tr>
                    <!---->
                    <!---->
                  </tbody>
                  <!---->
                </table>
                <!---->
              </div>
              <!---->
            </div>
             
            <button
              class="btn hide-target btn-outline-info"
              type="button"
            >
              
      非表示
    
            </button>
          </div>
           
          <hr
            class="mt-2 mb-3"
          />
           
          <div>
            <div
              class="form-group form-row"
              role="group"
            >
              <label
                class="col-form-label col-sm-3"
                for="reservation_status_id"
              >
                 予約ステータス 
              </label>
              <div
                class="col-sm-9"
              >
                <!---->
                <select
                  class="form-control"
                  id="reservation_status_id"
                >
                  <!---->
                  <option
                    data-key="0"
                    value=""
                  >
                      
                  </option>
                  <option
                    data-key="1"
                    value="1"
                  >
                     未予約 
                  </option>
                  <option
                    data-key="2"
                    value="2"
                  >
                     電話予約済 
                  </option>
                  <option
                    data-key="3"
                    value="3"
                  >
                     書類提出済 
                  </option>
                  <option
                    data-key="4"
                    value="4"
                  >
                     書類控取得 
                  </option>
                  <option
                    data-key="5"
                    value="5"
                  >
                     料金支払済 
                  </option>
                  <option
                    data-key="6"
                    value="6"
                  >
                     会計精算済 
                  </option>
                  <option
                    data-key="7"
                    value="7"
                  >
                     キャンセル 
                  </option>
                </select>
                <!---->
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
        <footer
          class="modal-footer"
        >
          <button
            class="btn btn-save btn-primary"
            type="submit"
          >
            
      保存
    
          </button>
        </footer>
      </div>
    </div>
  </div>
  <div
    class="modal-backdrop fade"
  />
</div>
`;
